File: doc-examples/oracle-jdbc-example-java/src/main/java/example/AbstractBookRepository.java

package example;

import io.micronaut.data.jdbc.annotation.JdbcRepository;
import io.micronaut.data.jdbc.runtime.JdbcOperations;
import io.micronaut.data.model.query.builder.sql.Dialect;
import io.micronaut.data.repository.CrudRepository;

import jakarta.transaction.Transactional;
import java.sql.ResultSet;
import java.util.List;

@JdbcRepository(dialect = Dialect.ORACLE)
public abstract class AbstractBookRepository implements CrudRepository<Book, Long> {

    private final JdbcOperations jdbcOperations;

    public AbstractBookRepository(JdbcOperations jdbcOperations) {
        this.jdbcOperations = jdbcOperations;
    }

    /**
     * Finds by title.
     *
     * @param title The title
     * @return List of books with matching title
     */
    @Transactional
    public List<Book> findByTitle(String title) {
        String sql = "SELECT * FROM Book book WHERE book.title = ?";
        return jdbcOperations.prepareStatement(sql, statement -> {
            statement.setString(1, title);
            ResultSet resultSet = statement.executeQuery();
            return jdbcOperations.entityStream(resultSet, Book.class).toList();
        });
    }
}

File: doc-examples/jdbc-example-python/src/main/python/example/AbstractBookRepository.py
from jakarta.transaction import Transactional
from micronaut.data.jdbc.annotation import JdbcRepository
from micronaut.data.jdbc.runtime import JdbcOperations
from micronaut.data.model.query.builder.sql import Dialect
from micronaut.data.repository import CrudRepository

from example.Book import Book


@JdbcRepository(dialect=Dialect.H2)
class AbstractBookRepository(CrudRepository[Book, int]):

    def __init__(self, jdbc_operations: JdbcOperations):
        self.jdbc_operations = jdbc_operations

    @Transactional
    def findByTitle(self, title: str) -> list[Book]:
        sql = "SELECT * FROM Book AS book WHERE book.title = ?"

        def query(statement):
            statement.setString(1, title)
            result_set = statement.executeQuery()
            return self.jdbc_operations.entityStream(result_set, Book).toList()

        return self.jdbc_operations.prepareStatement(sql, query)
